pkg/deployment/reconcile/plan_builder_rebuild_outsynced_shards.go

```diff
@@ -0,0 +1,121 @@
+//
+
+package reconcile
+
+import (
+	"context"
+
+	api "github.com/arangodb/kube-arangodb/pkg/apis/deployment/v1"
+	"github.com/arangodb/kube-arangodb/pkg/deployment/actions"
+	"github.com/arangodb/kube-arangodb/pkg/deployment/features"
+	"github.com/arangodb/kube-arangodb/pkg/util/globals"
+	"github.com/arangodb/kube-arangodb/pkg/util/k8sutil"
+)
+
+var disableRebuildOutSyncedShards = false
+
+// createRotateOrUpgradePlan
+func (r *Reconciler) createRebuildOutSyncedPlan(ctx context.Context, apiObject k8sutil.APIObject,
+	spec api.DeploymentSpec, status api.DeploymentStatus,
+	context PlanBuilderContext) api.Plan {
+	var plan api.Plan
+
+	if !features.RebuildOutSyncedShards().Enabled() {
+		// RebuildOutSyncedShards feature is not enabled
+		return nil
+	}
+
+	if disableRebuildOutSyncedShards {
+		// RebuildOutSyncedShards has been used already
+		return nil
+	}
+
+	shardStatus, ok := context.ShardsInSyncMap()
+	if !ok {
+		r.log.Error("Unable to get shards status")
+		return nil
+	}
+	agencyState, ok := context.GetAgencyCache()
+	if !ok {
+		r.log.Error("Unable to get agency state")
+		return nil
+	}
+
+	outSyncedMembers := map[string]api.MemberStatus{}
+
+	members := map[string]api.MemberStatus{}
+	for _, m := range status.Members.AsListInGroup(api.ServerGroupDBServers) {
+		members[m.Member.ID] = m.Member
+	}
+
+	// Get shards which are out-synced for more than defined timeout
+	outSyncedShardsIDs := shardStatus.NotInSyncSince(globals.GetGlobalTimeouts().ShardRebuild().Get())
+
+	if len(outSyncedShardsIDs) > 0 {
+		// Create plan for out-synced shards
+		for _, shardID := range outSyncedShardsIDs {
+			shard, exist := agencyState.GetShardDetailsByID(shardID)
+			if !exist {
+				r.log.Error("Shard servers not found", shardID, shard.Database)
+				continue
+			}
+
+			for _, server := range shard.Servers {
+				member, ok := members[string(server)]
+				if !ok {
+					r.log.Error("Member not found - we can not fix out-synced shard!", server)
+				} else {
+					r.log.Info("Shard is out-synced and its Tree will be rebuild", shardID, shard.Database, shard.Collection, member.ID)
+					outSyncedMembers[member.ID] = member
+
+					action := actions.NewAction(api.ActionTypeRebuildOutSyncedShards, api.ServerGroupDBServers, member).
+						AddParam("shardID", shardID).
+						AddParam("database", shard.Database)
+
+					plan = append(plan, action)
+				}
+			}
+		}
+
+		// Update member conditions
+		for _, member := range members {
+			shouldUpdate := false
+			if _, ok := outSyncedMembers[member.ID]; ok {
+				shouldUpdate = member.Conditions.Update(api.ConditionTypeOutSyncedShards, true, "Member has out-synced shard(s)", "")
+			} else {
+				shouldUpdate = member.Conditions.Remove(api.ConditionTypeOutSyncedShards)
+			}
+
+			if shouldUpdate {
+				if err := context.UpdateMember(ctx, member); err != nil {
+					r.log.Error("Can not save member condition", member.ID, api.ConditionTypeOutSyncedShards, err)
```

Comment:
lets say err is nil here, so member is not updated, but below variable `disableRebuildOutSyncedShards` will  be set to true, and this action will not be launched. Is it intended?

Reply:
Right now yes - we are not using that condition anywhere